File: atcoder/abc/abc097/d/main.py
class UnionFind:
	def __init__(self, N):
		self.root = list(range(N + 1))
		self.size = [1] * (N + 1)
		self.N = N
 
	def find_root(self, x):
		root = self.root
		while root[x] != x:
			root[x] = root[root[x]]
			x = root[x]
		return x
 
	def merge(self, x, y):
		root, size = self.root, self.size
		x = self.find_root(x)
		y = self.find_root(y)
		if x == y:
			return
		if size[x] < size[y]:
			x, y = y, x
		root[y] = root[x]
		size[x] += size[y]
	
	def same(self, x, y):
		return self.find_root(x)==self.find_root(y)

	def groups(self):
		cnt = set()
		seen = [False]*self.N
		for i in range(self.N+1):
			x = self.find_root(i)
			if not x==i: cnt.add(x)
		return len(cnt)

	def groups(self):
		G = [[] for _ in range(self.N)]
		for i in range(self.N):
			G[self.find_root(i)].append(i)
		return [i for i in G if i]

N, M = map(int, input().split())
uf =  UnionFind(N)
P = [int(i)-1 for i in input().split()]
for i in range(M):
  u, v = map(int, input().split())
  u -= 1
  v -= 1
  uf.merge(u, v)
pos = [-1]*N
for i, p in enumerate(P):
  pos[p] = i

ans = 0
for G in uf.groups():
  count = set()
  for x in G:
    count.add(pos[x])
  
  for x in G:
    if x in count:
      ans += 1
print(ans)

File: atcoder/abc/abc097/d/main.cpp
#include<bits/stdc++.h>
using namespace std;
struct dsu{
  int N;
  vector<int> root, size;
    dsu(int _N=0): root(_N), size(_N){
    for(int i=0;i<_N;i++){
      root[i]=i;
    }
    N = _N;
  }

  int merge(int x, int y){
    x = find(x);
    y = find(y);
    if (x==y) return -1;
    if (size[y]>size[x]) swap(x, y);
    root[y] = root[x];
    size[y] += size[x];
    return root[y];
  }

  bool same(int x, int y){
    return find(x)==find(y);
  }

  int find(int x){
    if (root[x]==x) return x;
    root[x] = find(root[x]);
    return root[x];
    
  }

  vector<vector<int>> groups(){
    vector<vector<int>> keep(N, vector<int>());
    vector<vector<int>> ret;
    for(int i=0;i<N;i++){
      keep[find(i)].push_back(i);
    }
    for(int i=0;i<N;i++){
      if(keep[i].size()!=0){
        ret.push_back(keep[i]);
      }
    }
    return ret;
  }
};

int main(){
  int N, M;
  cin >> N >> M;
  int u, v;
  dsu uf(N);
  vector<int> P(N);
  int p;
  for(int i=0;i<N;i++){
    cin >> p;
    P[p] = i;
  }
  for(int i=0;i<M;i++){
    cin >> u >> v;
    u--;
    v--;
    uf.merge(u, v); 
  }

  vector<int> V;
  int ans = 0;
  for(auto g: uf.groups()){
    V.clear();
    unordered_set<int> pos;
    for(auto x: g){
      V.push_back(x);
      pos.insert(x);
    }
    for(auto v: V) ans += pos.count(v);
  }
  cout << ans << endl;
}